pkg/consuldp/consul_dataplane.go

```diff
@@ -152,13 +154,25 @@ func (cdp *ConsulDataplane) Run(ctx context.Context) error {
 		return fmt.Errorf("failed to run proxy: %w", err)
 	}
 
+	dnsServer, err := cdp.startDNSProxy(ctx)
+	if err != nil {
+		if err != dns.ErrServerDisabled {
```

Comment:
If there is an error on start up, the envoy proxy process will linger around (as it runs as a sub-process) although the parent consul-dataplane process has died. 
It might be a good idea to start the DNS proxy before the envoy proxy so that if there is an issue on start-up, we hard fail right away before trying to start the envoy proxy as a sub-process.

Reply:
Hmm this seems to be more of a bug in main where we don't call cancel on the context if run returns a failure. I think it would make more sense to call cancel in the error handling in main. Thoughts?

Reply:
So far run returning a failure has been more immediate as a result of synchronous calls erroring out. 
If there is a failure in async go routines, [appropriate cleanup](https://github.com/hashicorp/consul-dataplane/blob/main/pkg/consuldp/consul_dataplane.go#L156-L172) is done before exiting. So if you are referring to calling `cancel` here - https://github.com/hashicorp/consul-dataplane/blob/main/cmd/consul-dataplane/main.go#L210, it might not work out since we would need to wait for everything to be cancelled before exiting?

Reply:
We should chat as that's not how I think we handle exiting. If there is an error the context should close and all dependent go routines should close. I did the cancel in that error handling and the behaviour you saw was fixed. 

I have tested it and it does work. The context handler stops everything and since it's a channel the communication should be async. It's the same behaviour as if you called SIGTERM.

Reply:
Sure, lets chat. Basically, there is nothing sending a stop to the envoy proxy explicitly since it does not share the parent process context.

Reply:
>It's the same behaviour as if you called SIGTERM.

It is slightly different because there is an async go routine listening for the cancel to handle appropriate cleanup: https://github.com/hashicorp/consul-dataplane/blob/main/pkg/consuldp/consul_dataplane.go#L156